python-lib/text_extraction/__init__.py

```diff
@@ -54,4 +54,137 @@ def extract_text_content(file_bytes, extension, with_pandoc):
             return file_bytes.decode()
         
         return text
-            
\ No newline at end of file
+
+
+def extract_text_chunks(filename, file_bytes, extension, with_pandoc):
+    if extension == "pdf":
+        pdf_pages = pdfium.PdfDocument(file_bytes)
+        return [
+            {
+                'file': filename,
+                'text': page.get_textpage().get_text_range(),
+                'id': page_id + 1,
+                'metadata': {"page": page_id + 1},
+                'error_message': ""
+            }
+            for page_id, page in enumerate(pdf_pages)
+        ]
+    elif extension == "doc":
+        raise ValueError("'doc' files are not supported, try to convert them to docx.")
+    elif extension == "md":
+        return extract_markdown_chunks(file_bytes.decode(), filename)
+    else:
+        if not with_pandoc:
+            raise ValueError("pandoc is required to extract chunks from files (except for PDFs and markdown).")
+
+        try:
+            temporary_job_folder = os.getcwd()
+            with tempfile.NamedTemporaryFile(dir=temporary_job_folder, suffix=".{}".format(extension)) as tmp:
+                tmp.write(file_bytes)
+                markdown = pypandoc.convert_file(tmp.name, to="md", format=extension)
+        except Exception as e:
+            raise ValueError("Cannot convert file into markdown using pandoc because: {}".format(e))
+    
+        return extract_markdown_chunks(markdown, filename)
+
+
+def extract_markdown_chunks(markdown, filename):
+    """
+    Extracts chunks from a markdown document. 
+    
+    The document is chunked according to its headers. Headers with levels 4+ are considered
+    as part of the text.
+
+    Returns a list of dictionaries, where each dictionary contains a chunk data and 
+    corresponding metadata.
+
+    This code is largely inspired by `langchain.text_splitter.MarkdownHeaderTextSplitter.split_text`
+    https://github.com/langchain-ai/langchain/blob/v0.0.333/libs/langchain/langchain/text_splitter.py#L376
+    """
+
+    lines = markdown.split("\n")
+    # Final output
+    lines_with_metadata = []
+    # text and metadata of the chunk currently being processed
+    current_text = []
+    current_metadata = {}
+    # Keep track of the nested header structure
+    # header_stack: List[Dict[str, Union[int, str]]] = []
+    header_stack = []
+    initial_metadata = {}
+
+    in_code_block = False
+
+    for line in lines:
+        stripped_line = line
+        if not stripped_line.startswith(("    ", "\t")):
+            # Do not remove tab or 4-spaces as they signal code blocks
+            # Only lines that start with less than 4 spaces can be header or code blocks
```

Comment:
I think we can focus on headers in the comment, to make things clearer. Mostly because lines starting with tab/4-spaces are not **necessarily** a code block

```suggestion
            # Header lines can only start with 0 to 3 spaces.
```